File: Mediator/Ruby/Command.rb
class Command
    def initialize(mediator, file)
        @mediator = mediator
        @file = file
    end

    def describe
        raise NotImplementedError
    end

    def execute
        if @mediator.checkStatus
            puts "Command not executed"
        else
            describe
            @mediator.lock(self)
            puts "Command executed"
        end
    end

    def finish
        @mediator.unlock
    end
end

File: Mediator/Javascript/Command.js
class Command {
  constructor(mediator, file) {
    this.mediator = mediator;
    this.file = file;
  }

  describe() {
    throw new Error("method not implemented");
  }

  execute() {
    if (this.mediator.checkStatus()) {
      console.log("Command not executed");
    } else {
      this.describe();
      this.mediator.lock(this);
      console.log("Command executed");
    }
  }

  finish() {
    this.mediator.unlock();
  }
}

export default Command;